s3transfer/futures.py

```diff
@@ -0,0 +1,246 @@
+        with self._lock:
+            if self._exception:
+                raise self._exception
+            return self._result
+
+    def cancel(self):
+        """Cancels the TransferFuture"""
+        with self._lock:
+            self._exception = futures.CancelledError
+            self._status = 'cancelled'
+
+    def set_status_to_running(self):
+        """Sets the TransferFuture's status to running"""
+        with self._lock:
+            self._status = 'running'
+
+    def done(self):
+        """Determines if a TransferFuture has completed
+
+        :returns: False if status is equal to 'failed', 'cancelled', or
+            'success'. True, otherwise
+        """
+        return self.status in ['failed', 'cancelled', 'success']
+
+    def add_failure_cleanup(self, function, *args, **kwargs):
+        """Adds a callback to call upon failure"""
+        with self._lock:
+            self._failure_cleanups.append(
+                functools.partial(function, *args, **kwargs)
+            )
+
+    def announce_done(self):
+        """Announce that future is done running
+
+        This allows the result() to be unblocked.
+        """
+        self._done_event.set()
+
+
+class BoundedExecutor(object):
+    EXECUTOR_CLS = futures.ThreadPoolExecutor
+
+    def __init__(self, max_size, max_num_threads):
+        """An executor implentation that has a maximum queued up tasks
+
+        The executor will block if the number of tasks that have been
+        submitted and is currently working on is past its maximum.
+
+        :params max_size: The maximum number of inflight futures. An inflight
+            future means that the task is either queued up or is currently
+            being executed. A size of None or 0 means that the executor will
+            have no bound in terms of the number of inflight futures.
+
+        :params max_num_threads: The maximum number of threads the executor
+            uses.
+        """
+        self._max_size = max_size
+        self._max_num_threads = max_num_threads
+        self._executor = self.EXECUTOR_CLS(max_workers=self._max_num_threads)
+        self._currently_running_futures = set()
+        self._lock = threading.Lock()
+
+    def submit(self, fn, *args, **kwargs):
+        """Submit a task to complete
+
+        :param fn: The function to call
+        :param args: The positional arguments to use to call the function
+        :param kwargs: The keyword arguments to use to call the function
+
+        :returns: The future assocaited to the submitted task
+        """
+        with self._lock:
+            # First determine if there are already too many futures running
+            if self._at_maximum_running_futures():
+                # If there are too many futures running, wait till some
+                # complete and save the remaining running futures as the
+                # next set to wait for.
+                self._currently_running_futures = futures.wait(
```

Comment:
Do you know if this is creating a copy of the incomplete waiters each time?

I'm wondering about the case where the input rate is greater than the output rate which will mean we're essentially calling `.wait()` every time we try to submit.  If the max size is large enough, and we are creating a copy of incomplete futures every `wait()` call, this seems like this could create a lot of overhead.

Thoughts?